packages/program/src/utils.ts

```diff
@@ -38,3 +48,96 @@ export function getAbisFromAllCalls(
     return acc;
   }, {} as JsonAbisFromAllCalls);
 }
+
+/** @hidden */
+export const extractInvocationResult = <T>(
+  functionScopes: Array<InvocationScopeLike>,
+  receipts: TransactionResultReceipt[],
+  isMultiCall: boolean,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  logs: any[]
+) => {
+  const mainCallConfig = functionScopes[0]?.getCallConfig();
+
+  if (functionScopes.length === 1 && mainCallConfig && 'bytes' in mainCallConfig.program) {
+    return callResultToInvocationResult<T>({ receipts }, mainCallConfig, logs);
+  }
+  const encodedResults = decodeContractCallScriptResult(
+    { receipts },
+    (mainCallConfig?.program as AbstractContract).id,
+    logs
+  );
+
+  const decodedResults = encodedResults.map((encodedResult, i) => {
+    const { func } = functionScopes[i].getCallConfig();
+    return func.decodeOutput(encodedResult)?.[0];
+  });
+
+  return (isMultiCall ? decodedResults : decodedResults?.[0]) as T;
+};
+
+type BuiltSubmitResultParams = {
+  funcScope: InvocationScopeLike | Array<InvocationScopeLike>;
+  transactionResponse: TransactionResponse;
+  isMultiCall: boolean;
+  program: AbstractProgram;
+};
+
+/** @hidden */
+export const buildSubmitResult = async <T>(
+  params: BuiltSubmitResultParams
+): Promise<SubmitResult<T>> => {
+  const { funcScope, isMultiCall, program, transactionResponse } = params;
+  const txResult = await transactionResponse.waitForResult();
+  const { receipts } = txResult;
+
+  const functionScopes = Array.isArray(funcScope) ? funcScope : [funcScope];
+  const mainCallConfig = functionScopes[0]?.getCallConfig();
+
+  const { main, otherContractsAbis } = getAbisFromAllCalls(functionScopes);
+  const logs = mainCallConfig ? getDecodedLogs(receipts, main, otherContractsAbis) : [];
+  const value = extractInvocationResult<T>(functionScopes, receipts, isMultiCall, logs);
+  const gasUsed = getGasUsedFromReceipts(receipts);
+
+  const submitResult: SubmitResult<T> = {
+    isMultiCall,
+    functionScopes,
+    value,
+    program,
+    transactionResult: txResult as TransactionResult<TransactionType.Script>,
+    transactionResponse,
+    transactionId: transactionResponse.id,
+    logs,
+    gasUsed,
+  };
+
+  return submitResult;
+};
+
+type BuiltDryRunResultParams = {
+  funcScopes: InvocationScopeLike | Array<InvocationScopeLike>;
+  callResult: CallResult;
+  isMultiCall: boolean;
+};
+
+/** @hidden * */
+export const buildDryRunResult = <T>(params: BuiltDryRunResultParams): DryRunResult<T> => {
+  const { funcScopes, callResult, isMultiCall } = params;
```

Comment:
Since we already refactoring to improve the readability of this code, we should have a consistent series of inline functions or destructures , this will also reuse some of the code. e.g.

```ts
  const { funcScopes, callResult, isMultiCall } = params;
  const functionScopes = getFunctionScopes(funcScopes);
  const { receipts } = callResult;
  const mainCallConfig = getMainCallConfig(functionScopes);
  const logs = getLogs(receipts, mainCallConfig, functionScopes);
  const value = extractInvocationResult<T>(functionScopes, receipts, isMultiCall, logs);
  const gasUsed = getGasUsedFromReceipts(receipts);

  return {
    functionScopes,
    callResult,
    isMultiCall,
    gasUsed,
    value,
  };
  ```
  
  We can create a series of re-usable utility functions
  
  ```ts
  const getLogs = (
  receipts: TransactionResultReceipt[],
  mainCallConfig: CallConfig | undefined,
  functionScopes: Array<InvocationScopeLike>
) => {
  if (!mainCallConfig) return [];
  const { main, otherContractsAbis } = getAbisFromAllCalls(functionScopes);
  return getDecodedLogs(receipts, main, otherContractsAbis);
};
```

```ts
const getFunctionScopes = (funcScope: InvocationScopeLike | Array<InvocationScopeLike>) =>
  Array.isArray(funcScope) ? funcScope : [funcScope];
```  

```ts
const getMainCallConfig = (functionScopes: Array<InvocationScopeLike>) =>
  functionScopes[0]?.getCallConfig();
  ```

We can then reuse these same functions in `buildSubmitResult`

```ts
export const buildSubmitResult = async <T>(
  params: BuiltSubmitResultParams
): Promise<SubmitResult<T>> => {
  const { funcScope, isMultiCall, program, transactionResponse } = params;
  const txResult = await transactionResponse.waitForResult();
  const { receipts } = txResult;
  const functionScopes = getFunctionScopes(funcScope);
  const mainCallConfig = getMainCallConfig(functionScopes);
  const logs = getLogs(receipts, mainCallConfig, functionScopes);
  const value = extractInvocationResult<T>(functionScopes, receipts, isMultiCall, logs);
  const gasUsed = getGasUsedFromReceipts(receipts);

  return {
    isMultiCall,
    functionScopes,
    value,
    program,
    transactionResult: txResult as TransactionResult<TransactionType.Script>,
    transactionResponse,
    transactionId: transactionResponse.id,
    logs,
    gasUsed,
  };
};
```

Reply:
https://github.com/FuelLabs/fuels-ts/pull/2652/commits/004f3f2379500858ef02d242b5027494b7a65b7d